mijia/src/bluetooth.rs

```diff
@@ -0,0 +1,201 @@
+        let dbus_handle = tokio::spawn(async {
+            let err = dbus_resource.await;
+            // TODO: work out why this err isn't 'static and use anyhow::Error::new instead
+            Err::<(), anyhow::Error>(anyhow::anyhow!(err))
+        });
+        Ok((
+            dbus_handle.map(|res| Ok(res??)),
+            BluetoothSession { connection },
+        ))
+    }
+
+    /// Power on the bluetooth adapter adapter and start scanning for devices.
+    pub async fn start_discovery(&self) -> Result<(), anyhow::Error> {
+        let adapter = dbus::nonblock::Proxy::new(
+            "org.bluez",
+            "/org/bluez/hci0",
+            DBUS_METHOD_CALL_TIMEOUT,
+            self.connection.clone(),
+        );
+        adapter
+            .set_powered(true)
+            .await
+            .with_context(|| std::line!().to_string())?;
+        adapter
+            .start_discovery()
+            .await
+            .unwrap_or_else(|err| println!("starting discovery failed {:?}", err));
+        Ok(())
+    }
+
+    /// Get a list of all bluetooth devices which have been discovered so far.
+    pub async fn get_devices(&self) -> Result<Vec<DeviceInfo>, anyhow::Error> {
+        let bluez_root = dbus::nonblock::Proxy::new(
+            "org.bluez",
+            "/",
+            DBUS_METHOD_CALL_TIMEOUT,
+            self.connection.clone(),
+        );
+        let tree = bluez_root.get_managed_objects().await?;
+
+        let sensors = tree
+            .into_iter()
+            .filter_map(|(path, interfaces)| {
+                // FIXME: can we generate a strongly typed deserialiser for this,
+                // based on the introspection data?
+                let device_properties = interfaces.get("org.bluez.Device1")?;
+
+                let mac_address = device_properties
+                    .get("Address")?
+                    .as_iter()?
+                    .filter_map(|addr| addr.as_str())
+                    .next()?
+                    .to_string();
+                // UUIDs don't get populated until we connect. Use:
+                //     "ServiceData": Variant(InternalDict { data: [
+                //         ("0000fe95-0000-1000-8000-00805f9b34fb", Variant([48, 88, 91, 5, 1, 23, 33, 215, 56, 193, 164, 40, 1, 0])
+                //     )], outer_sig: Signature("a{sv}") })
+                // instead.
+                let service_data: HashMap<String, Vec<u8>> = device_properties
+                    .get("ServiceData")?
+                    // Variant(...)
+                    .as_iter()?
+                    .next()?
+                    // InternalDict(...)
+                    .as_iter()?
+                    .tuples::<(_, _)>()
+                    .filter_map(|(k, v)| {
+                        let k = k.as_str()?.into();
+                        let v = v.box_clone();
+                        let v = cast::<Vec<u8>>(&v)?.clone();
```

Comment:
I don't trust that this cast call is going to work when it encounters real data. I would be more comfortable if we had a test. Maybe deploy it to cottagepi and see if it works before merging?

I suppose I should have made a test for this as well.

Reply:
I put it on cottagepi with some debugging and debugs as `Variant(InternalArray { data: [48, 88, 91, 5, 1, 23, 33, 215, 56, 193, 164, 40, 1, 0], inner_sig: Signature("y") })` but then casts to None.